### PR TITLE
[feat] AdaScale: Gradient Accumulation and Add PyTest unit tests

### LINKED ISSUE #146: pytest AdaScale
## 🚀 Feature

Once Mike and I gain more experiences with AdaScale with different projects, we should get it pytested. Before that, we will not use it widely in different projects.

## Motivation

Make sure code is covered with unit test.

## Pitch

pretty standard.

## Alternatives

no test?

## Additional context

Mike and I are doing some reproducibility study with AdaScale.

cc: @msbaines @blefaudeux

### LINKED ISSUE #195: AdaScale: work with gradient accumulation
Mike reported that this doesn't seem to work well. I will debug it. In the end, I will ensure it works and tested.

### PR DESCRIPTION
**Pull request by min-xu-ai**
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [x] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/master/CONTRIBUTING.md)?
- [x] Did you make sure to update the docs?
- [x] Did you write any new necessary tests?

## What does this PR do?
Fixes #195 .
Fixes #146 .

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by blefaudeux** [comment]
to take a step back, it could also be that we're not precise enough with mean and std, these are asymptotically good estimators if (a) enough samples and (b) single normal/gaussian distribution, a or b could be faulty here (for instance if other jobs are scheduled from time to time on the same hosts, not sure how that works). More robust estimators are for instance median and MAD (median of absolute differences), equivalents to mean/std (first and second moment estimators) but not pulled out by outliers -at the cost of slower convergence if your samples are from a gaussian actually-, I could move the regression test to that for instance ?

**Comment by blefaudeux** [comment]
@min-xu-ai is that still a draft on purpose, or is that open for review ?

**Comment by blefaudeux** [comment]
> to take a step back, it could also be that we're not precise enough with mean and std, these are asymptotically good estimators if (a) enough samples and (b) single normal/gaussian distribution, a or b could be faulty here (for instance if other jobs are scheduled from time to time on the same hosts, not sure how that works). More robust estimators are for instance median and MAD (median of absolute differences), equivalents to mean/std (first and second moment estimators) but not pulled out by outliers -at the cost of slower convergence if your samples are from a gaussian actually-, I could move the regression test to that for instance ?

commented on the phone and missed the proper PR, this message was intended to #203, sorry for the mixup...

**Comment by min-xu-ai** [comment]
> 
> 
> @min-xu-ai is that still a draft on purpose, or is that open for review ?

Thanks Ben! Yeah, I still need to add a few tests before making PR ready for review. Thanks for checking on it!

**Comment by min-xu-ai** [comment]
Hi guys, sorry this took a while. I am happy with the current state of the tests now. Let me know what do you think! Thanks a bunch!

File: tests/optim/test_single_node_adascale.py [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+
+
+@skip_if_no_gpu
+def test_grad_accum_gpu():
+    """Test the basic functionality on GPU with gradient accumulation without DDP"""
+    test_grad_accum_cpu(cpu=False)
+
+
+@skip_if_no_gpu
+def test_state_checkpointing():
```
Comment (mikerabbat) [inline tests/optim/test_single_node_adascale.py:94]: Nice! Thanks for remembering this

**Review by mikerabbat — APPROVED** [review]
Looks good to me! Just a few minor typos and suggestions

File: README.md [hunk 2]
```diff
@@ -12,6 +12,7 @@ fairscale supports:
    * tensor parallelism (fairscale.nn.model_parallel)
 * Optimization:
    * optimizer state sharding (fairscale.optim.oss)
+   * AdaScale SGD (fairscale.optim.AdaScale)
```
[the remark this answers is not in the document]
Comment (min-xu-ai) [reply README.md:15]: You are right. How about I change it to `from fairscale.optim import AdaScale`?

File: README.md [hunk 3]
```diff
@@ -103,7 +104,18 @@ if __name__ == "__main__":
...
 
+AdaScale can be used to wrap a SGD optimizer and to be used in DDP (Distributed Data Parallel)
+training or non-DDP with gradient accumulation. The benefit is to re-use the same LR
+schedule from a baseline batch size when effective batch size is bigger.
+
+Primary goal is to allow scaling to bigger batch sizes without losing model accuracy.
+
+At a high level, we want ML researchers to:
+  * go parallel more easily (i.e. reuse the same LR scahedule)
+  * worry not about accuracy loss
```
[the remark this answers is not in the document]
Comment (min-xu-ai) [reply README.md:117]: Cool. Will do!

File: README.md [hunk 4]
```diff
@@ -103,7 +104,18 @@ if __name__ == "__main__":
...
+### AdaScale SGD
 
+AdaScale can be used to wrap a SGD optimizer and to be used in DDP (Distributed Data Parallel)
+training or non-DDP with gradient accumulation. The benefit is to re-use the same LR
+schedule from a baseline batch size when effective batch size is bigger.
+
+Primary goal is to allow scaling to bigger batch sizes without losing model accuracy.
+
+At a high level, we want ML researchers to:
+  * go parallel more easily (i.e. reuse the same LR scahedule)
```
[the remark this answers is not in the document]
Comment (min-xu-ai) [reply README.md:116]: great catch. I was using a different keyboard yesterday. :-) fixed.

File: fairscale/optim/adascale.py [hunk 5]
```diff
@@ -65,14 +66,15 @@ class AdaScale(object):
...
             Scaling factor of the batch size, e.g. using a 10x
             larger batch size (summed across all world_size) means a scale of
             10. If None, defaults to ``world_size``.
-        patch_optimizer (bool):
-            If True, monkey-patches the ``step`` method of
-            the optimizer with the AdaScale's ``step`` method.
+        smoothing (float):
+            Smoothing factor between batches. Default value: 0.9999
+        num_gradients_to_accumulate (int):
+            Number of passes that we accumulate gradients locally.
```
[the remark this answers is not in the document]
Comment (min-xu-ai) [reply fairscale/optim/adascale.py:77]: great idea. done.

File: fairscale/optim/adascale.py [hunk 6]
```diff
@@ -208,63 +244,98 @@ def _queue_callback(self) -> None:
...
-        self._local_grad_sqr.div_(self._world_size)
+        # Keep track of number of backward calls for gradient accumulation.
+        self._num_backward_calls += 1
 
-        local_grad_sqr = self._local_grad_sqr.cpu().numpy()
+        # TODO (min): We need to have a way to check that training loop & DDP
+        #             is doing the right thing where the gradient is reduced
+        #             in this backward pass.
+        #             Longer term, we may compute the gain and then inform
+        #             the training loop when it is a good time to step().
```
[the remark this answers is not in the document]
Comment (min-xu-ai) [reply fairscale/optim/adascale.py:260]: Thanks! I will put our names there then. :-)

File: fairscale/optim/adascale.py [hunk 7]
```diff
@@ -208,63 +244,98 @@ def _queue_callback(self) -> None:
...
-        self._local_grad_sqr = None
 
     def step(self, *args: Any, **kwargs: Any) -> Optional[float]:
         """
         Run one optimizer step using Adascale. Essentially just invokes
         ``optimizer.step(*args, **kwargs)`` with a scaled learning rate.
 
         Args:
-            args:
+            args (Any):
```
[the remark this answers is not in the document]
Comment (min-xu-ai) [reply fairscale/optim/adascale.py:307]: I agree with you. I think sticking to pytorch interface is better. I don't know why @aurickq had it this way to begin with.

File: tests/optim/test_ddp_adascale.py [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+def _dist_init(rank, world_size, tempfile_name, backend):
+    url = "file://" + tempfile_name
+    dist.init_process_group(init_method=url, backend=backend, rank=rank, world_size=world_size)
+    torch.cuda.set_device(rank)
+
+
+def _test_basic_func(rank, world_size, tempfile_name):
+    _dist_init(rank, world_size, tempfile_name, backend="nccl")  # Covers nccl
+
+    model = Linear(2, 2, bias=False)
```
[the remark this answers is not in the document]
Comment (min-xu-ai) [reply tests/optim/test_ddp_adascale.py:38]: nice example! I initially had weights assigned but then I realized that the gain is anyway deterministic in the way I test it. So I removed it.

File: tests/optim/test_single_node_adascale.py (hunk 1 above)
Comment (min-xu-ai) [reply tests/optim/test_single_node_adascale.py:94]: Yeah, it bugs me a bit too. :-)

File: CHANGELOG.md [hunk 9]
```diff
@@ -5,6 +5,13 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
 
+## [next rel] - TBD
```
Comment (blefaudeux) [inline CHANGELOG.md:8]: oh good idea, I should do that also

**Review by blefaudeux — APPROVED** [review]
LGTM, thanks for the small but important changes !

File: fairscale/optim/adascale.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (mikerabbat) [reply fairscale/optim/adascale.py:308]: Hmm... maybe better to leave it for now actually. I was just looking through the fairseq code, and they take an additional argument.https://fairseq.readthedocs.io/en/latest/_modules/fairseq/optim/fairseq_optimizer.html#FairseqOptimizer.step

[the remark this answers is not in the document]
Comment (min-xu-ai) [reply fairscale/optim/adascale.py:308]: Great find! Will leave it as is for now.
